src/vs/editor/contrib/referenceSearch/browser/referencesModel.ts

```diff
@@ -201,9 +208,20 @@ export class ReferencesModel implements IDisposable {
 	private _references: OneReference[] = [];
 	private _eventBus = new EventEmitter();
 
-	onDidChangeReferenceRange: Event<OneReference> = fromEventEmitter<OneReference>(this._eventBus, 'ref/changed');
+	onDidChangeReferenceRange: Event<OneReference> = fromEventEmitter<OneReference>(this._eventBus, 'range/changed');
+	onDidChangeReferences: Event<OneReference> = fromEventEmitter<OneReference>(this._eventBus, 'refs/changed');
 
 	constructor(references: Location[]) {
+		this._setReferences(references, false);
+	}
+
+	public setReferences(references: Location[]) {
+		this._setReferences(references, true);
```

Comment:
I think this should be an append-style method, so instead of resetting and resorting all locations we should always have append-insert semantics. So, if provider A returned `A` and `C` and if slow provider B returns `B` the order should be `A`, `C`, `B`. My goal is a stable UI, think of the following: `A` and `C` are in the UI and the user is just hovering over `C` to select it. At that time  `B` is getting resolved. It should not push down `C` because the user would then select `B` instead of `C`. The tricky bit is that we are talking about a tree. I not sure what to do when the tree contains `A/B` and `C/D`, the user is selecting `D` and at the same time another child is inserted into the `A`-subtree. . It will push down the `C`-subtree, resulting in an unstable-ui.